lib/spack/spack/solver/asp.py

```diff
@@ -3723,6 +3832,102 @@ def _order_index(flag_group):
+        splice_node: NodeArgument,
+        child_name: str,
+        child_hash: str,
+    ):
+        splice = Splice(splice_node, child_name=child_name, child_hash=child_hash)
+        if parent_node not in self._splices:
+            self._splices[parent_node] = []
+        self._splices[parent_node].append(splice)
+
+    def splice_in_dependency(
+        self, parent_node: NodeArgument, child_node: Union[str, NodeArgument]
+    ):
+        if parent_node not in self._splice_dag:
+            self._splice_dag[parent_node] = []
+        # Hash is the base-case covered by splice_at_hash
+        if isinstance(child_node, NodeArgument):
+            self._splice_dag[parent_node].append(child_node)
+
+    def _resolve_splices_for_node(
+        self, node: NodeArgument, resolved: Dict[NodeArgument, spack.spec.Spec]
+    ) -> spack.spec.Spec:
+        """
+        This function both caches it's answer in resolved, and returns the
+        spec resulting from resolving splices in the nodes
+        """
+        # Bottom up caching
+        if node in resolved:
+            return resolved[node]
+        orig_spec = self._specs[node]
+        immediate_splices = self._splices.get(node, [])
+        deps_with_splices = self._splice_dag.get(node, [])
+        # This node has no splicing to be done to it.
+        if len(immediate_splices) == 0 and len(deps_with_splices) == 0:
+            resolved[node] = orig_spec
+            return orig_spec
+        new_spec = orig_spec.copy(deps=False)
+        new_spec.clear_cached_hashes((ht.package_hash.attr,))
+        edges_by_dep_name: Dict[str, List[spack.spec.DependencySpec]]
+        edges_by_dep_name = {}
+        for edge in orig_spec.edges_to_dependencies():
+            edge_name = edge.spec.name
+            assert type(edge_name) is str, "Anonymous dependency spec in splice"
+            if edge_name not in edges_by_dep_name:
+                edges_by_dep_name[edge_name] = []
+            edges_by_dep_name[edge_name].append(edge)
+        # This is the easy case, we can just copy unspliced dependencies
+        for name, edges in edges_by_dep_name.items():
+            if name not in [s.child_name for s in immediate_splices] and name not in [
+                n.pkg for n in deps_with_splices
+            ]:
+                for e in edges:
+                    new_spec.add_dependency_edge(e.spec, depflag=e.depflag, virtuals=e.virtuals)
+        # This is the case where splices may occurr deeper in dependencies
+        potential_clean_edges = set()
+        dirty_edges = set()
+        for dep_node in deps_with_splices:
+            old_dep_spec = self._specs[dep_node]
+            dep_name = old_dep_spec.name
+            edges = edges_by_dep_name[dep_name]
+            for e in edges:
+                if e.spec.dag_hash() == old_dep_spec.dag_hash():
+                    new_dep_spec = self._resolve_splices_for_node(dep_node, resolved)
+                    other_deps = e.depflag & ~dt.BUILD
```

Comment:
This would be better named `depflag` or `non_build_deps` -- I think the name is vestigial from a time in the splicing PR that I was separating `build_deps` and `other_deps` into two separate flags, and I think a poorly named variable slipped into develop through that.